packages/amplication-server/src/core/resource/resource.service.ts

```diff
@@ -827,6 +860,123 @@ export class ResourceService {
+      where: {
+        projectId: project.id,
+        resourceType: EnumResourceType.Service,
+        deletedAt: null,
+      },
+    });
+    if (projectResources.length > 0) {
+      for (const newService of newServices) {
+        // duplicate name validation
+        const duplicateService = projectResources.find(
+          (resource) =>
+            resource.name.toLocaleLowerCase() ===
+            newService.name.toLocaleLowerCase()
+        );
+        if (duplicateService) {
+          throw new AmplicationError(
+            `Resource : ${newService.name} is already exist in project: ${project.name}.`
+          );
+        }
+        //service limitation validation
+        const featureServices = await this.billingService.getMeteredEntitlement(
+          project.workspaceId,
+          BillingFeature.Services
+        );
+
+        if (
+          !featureServices.hasAccess ||
+          (!featureServices.isUnlimited &&
+            featureServices.usageLimit <= projectResources.length)
+        ) {
+          throw new AmplicationError(SERVICE_LIMITATION_ERROR);
+        }
+      }
+    }
+  }
+  private async validateMovedEntitiesData(
+    movedEntitiesByResource: {
+      [resourceId: string]: RedesignProjectMovedEntity[];
+    },
+    project: Project,
+    originalResourceId: string,
+    user: User
+  ): Promise<void> {
+    // entities limitation per service validation
+    const featureEntitiesServices =
+      await this.billingService.getNumericEntitlement(
+        project.workspaceId,
+        BillingFeature.EntitiesPerService
+      );
+
+    const serviceSettings =
+      await this.serviceSettingsService.getServiceSettingsValues(
+        {
+          where: { id: originalResourceId },
+        },
+        user
+      );
+
+    for (const [resourceId, entities] of Object.entries(
+      movedEntitiesByResource
+    )) {
+      for (const movedEntity of entities) {
+        const currentEntity = await this.entityService.entity({
+          where: {
+            id: movedEntity.entityId,
+          },
+        });
+        // authEntity validation
+        if (
+          serviceSettings.authEntityName &&
+          serviceSettings.authEntityName === currentEntity.name
+        ) {
+          throw new AmplicationError(
+            `cannot move auth entity : ${currentEntity.name}.`
+          );
+        }
+
+        const resourceEntities = await this.entityService.entities({
+          where: {
+            resourceId: movedEntity.targetResourceId,
+            deletedAt: null,
+          },
+        });
+
+        //duplicate entities names validation
+        if (resourceEntities.length > 0) {
+          for (const resourceEntity of resourceEntities) {
+            if (resourceEntity.name === currentEntity.name) {
+              throw new AmplicationError(
+                `Entity : ${currentEntity.name} is already exist in resource: ${movedEntity.targetResourceId}.`
```

Comment:
Let's discuss if we want to bring the targetResource name here or not

Reply:
IIUC, targetResourceId is the same as resourceId at this point (in the map) and we already get the current resource in the next lines